alternator/error.hh

```diff
@@ -44,7 +54,24 @@ public:
     { }
     api_error() = default;
     virtual const char* what() const noexcept override { return _msg.c_str(); }
+protected:
+    api_error(detail::exception_type, std::string msg, status_type http_code);
 };
 
+namespace detail {
+template<detail::exception_type Type>
+class t_api_error : public api_error {
+public:
+    t_api_error(std::string msg, status_type http_code = status_type::bad_request)
+        : api_error(Type, std::move(msg), http_code)
+    {}
+};
+}
+
+using validation_exception = detail::t_api_error<detail::exception_type::ValidationException>;
```

Comment:
A factory function is the exact same thing, but C-style (and with a lot of boiler plate duplication for the various errors) and without the benefit of separate type (i.e. discernable on catch/log level). I fail to see how this is better. Seriously, the template here is literally the simplest and least error prone you can get: Cannot use the wrong type + all aspects handled by code completion/intellisense.

Reply:
I'll just have to send a followup patch with my suggestion. I think it is exactly the opposite of what you said - there will be less boiler-plate code than you had to add, not more, code will be clearer, simpler, and usages will be clearer - when you write api_error::something its clearer that it is a subclass of api_error than just something_exception which can be any C++ exception thing. 
 don't think that "C-style" is a good way to judge anything. There is no "opposite of C-style" - there is no C++ style. C++ style was once upon a time heavily based on OOP, today it's heavily generics (templates), and you actually used both techniques in a place that doesn't need either :-) When I'll send a followup patch you can see what I meant.

Reply:
Ok, you mean `static [noreturn] void api_error::throw_validation_exception(sstring msg)`. Fair enough, but that has the downside of a) not being able to _catch_ specific error. b) if the throw is in a factory function which is not inline thw compiler can miss optimization ops. 
Can of course use  `static api_error api_error::throw_validation_exception(sstring msg)`, but still lacks a)
Personally, I prefer specific types, since that is how exception handling was meant to be used.

Reply:
I will send a follow-up patch which undoes your templated hierarchy and returns a **single** type of exception, `api_error`, with multiple factory functions to build them (instead of explicit strings).
The idea of having a single type of `api_error` exception wasn't because I like C so much (?). It was because there is only one way to return an "api error" in the HTTP response - containing an error "name" and JSON content - and this is what "api_error" contains. Nobody is expected to "catch" the different types of api_errors - they are designed to be directly sent to the user, as is. This is why producers of these errors throw them. Code which throws exceptions which should be caught - should throw other types of exceptions - not api_errors (if such an exception goes all the way back to the user, we translate it to an "internal error" api_error).

Yes, what I want is something like `static api_error api_error::validation(sstring msg)`. Like this - no need for the extra long name "make_validation_exception" - the name `api_error::validation` is clear and shorter. 

Again, I'll do this as a followup patch and you can see what it looks like, and object to it (if you still do) then.